hypervisor/dm/vpci/msi.c

```diff
@@ -89,15 +79,39 @@ static int vmsi_remap(struct pci_vdev *vdev, bool enable)
 		pci_pdev_write_cfg(pbdf, capoff + PCIR_MSI_DATA, 0x2U, (uint16_t)info.pmsi_data);
 	}
 
-	/* If MSI Enable is being set, make sure INTxDIS bit is set */
-	if (enable) {
-		enable_disable_pci_intx(pbdf, false);
-		pci_pdev_write_cfg(pbdf, capoff + PCIR_MSI_CTRL, 2U, msgctrl | PCIM_MSICTRL_MSI_ENABLE);
+	return ret;
+}
+
+static int vmsi_enable(struct pci_vdev *vdev)
+{
+	union pci_bdf pbdf = vdev->pdev.bdf;
+	uint32_t msgoff = vdev->msi.capoff + PCIR_MSI_CTRL;
+	uint32_t msgctrl;
+	int ret = 0;
+
+	ret = vmsi_update(vdev);
```

Comment:
do we still need to call vmsi_update in vmsi_enable?

Reply:
Yes, currently vmsi_update() is called in 2 cases:
1) When MSI Enable bit is being set
2) Message Data/Addr is being changed when MSI is enabled

As Fei suggested, we can choose to remove 1) and in 2), we need to call vmsi_update() regardless MSI Enable is set or not. Similar logic applies to msix.c as well.

Reply:
Is it safe to call vmsi_update with MSI enable?
if it is not safe, even guest does it, HV should avoid it.